src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,829 @@
+        {
+            if (ShouldRemoveVariableDeclaratorContainingExpression(document, expression, out var varDeclName))
+            {
+                return varDeclName;
+            }
+
+            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var semanticFacts = document.GetRequiredLanguageService<ISemanticFactsService>();
+            return semanticFacts.GenerateNameForExpression(semanticModel, expression, capitalize: false, cancellationToken);
+        }
+
+        /// <summary>
+        /// Determines if the expression's grandparent is a variable declarator and if so,
+        /// returns the name
+        /// </summary>
+        protected static bool ShouldRemoveVariableDeclaratorContainingExpression(
+            Document document, TExpressionSyntax expression, [NotNullWhen(true)] out string? varDeclName)
+        {
+            var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+            var expressionDecl = expression?.Parent?.Parent;
+            if (syntaxFacts.IsVariableDeclarator(expressionDecl))
+            {
+                varDeclName = syntaxFacts.GetIdentifierOfVariableDeclarator(expressionDecl).ValueText;
+                return true;
+            }
+
+            varDeclName = null;
+            return false;
+        }
+
+        /// <summary>
+        /// Locates all the call sites of the method that introduced the parameter
+        /// </summary>
+        protected static async Task<ImmutableDictionary<Document, ImmutableArray<TInvocationExpressionSyntax>>> FindCallSitesAsync(
+            Document document, IMethodSymbol methodSymbol, CancellationToken cancellationToken)
+        {
+            var methodCallSitesBuilder = ImmutableDictionary.CreateBuilder<Document, ImmutableArray<TInvocationExpressionSyntax>>();
+            var progress = new StreamingProgressCollector();
+            await SymbolFinder.FindReferencesAsync(
+                methodSymbol, document.Project.Solution, progress: progress,
+                documents: null, FindReferencesSearchOptions.Default, cancellationToken).ConfigureAwait(false);
+            var referencedSymbols = progress.GetReferencedSymbols();
+
+            // Ordering by descending to sort invocations by starting span to account for nested invocations
+            var referencedLocations = referencedSymbols.SelectMany(referencedSymbol => referencedSymbol.Locations).Distinct()
+                .OrderByDescending(reference => reference.Location.SourceSpan.Start);
+            var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+
+            // Adding the original document to account for refactorings that do not have any invocations
+            methodCallSitesBuilder.Add(document, new ArrayBuilder<TInvocationExpressionSyntax>().ToImmutable());
```

Comment:
first, just to ImmutableArray<TINvocation>.Empty to get an empty array :)
second, this approach isn't awesome.  because you have an immutable array, every time you add to it, you will copy every item in teh array.  this makes it n^2 with a *lot* of allocations.

it's honestly fine to jsut make a Dictionary<Document, List<TInvocationExpression>>.

Reply:
changed back